### PR TITLE
CBG-166 Specify body copy policy on rev cache retrieval

### PR DESCRIPTION
**Pull request by adamcfraser**
Avoids duplicate copy during delta processing (shallow, then deep).

### CODE REVIEW

File: db/revision_cache.go [hunk 1]
```diff
@@ -127,7 +115,7 @@ func (rc *RevisionCache) GetActive(docid string, context *DatabaseContext) (docR
 
 	// Retrieve from or add to rev cache
 	value := rc.getValue(docid, bucketDoc.CurrentRev, true)
-	docRev, statEvent, err := value.loadForDoc(bucketDoc, context)
+	docRev, statEvent, err := value.loadForDoc(bucketDoc, context, BodyShallowCopy)
```
Comment (tleyden) [inline db/revision_cache.go:118]: I think this might have been my fault from another PR, but I think the function signature:

```
func (value *revCacheValue) load(loaderFunc RevisionCacheLoaderFunc, copyType BodyCopyType) (DocumentRevision, bool, error) {

```

should be changed at some point to have a named return type.  I think just having `bool` is a bit unclear, and puts the burden on the person reading the code to look at internal variable names or calling code to figure out the nature of the return val.

File: rest/blip_api_test.go [hunk 2]
```diff
@@ -1589,3 +1595,47 @@ func TestBlipDeltaSyncPush(t *testing.T) {
 	assert.Equal(t, http.StatusOK, resp.Code)
 	assert.Equal(t, `{"_id":"doc1","_rev":"2-abcxyz","greetings":[{"hello":"world!"},{"hi":"alice"},{"howdy":"bob"}]}`, resp.Body.String())
 }
+
+// TestBlipNonDeltaSyncPush tests that a client that doesn't support deltas can push to a SG that supports deltas (either CE or EE)
+func TestBlipNonDeltaSyncPush(t *testing.T) {
+
+	defer base.SetUpTestLogging(base.LevelTrace, base.KeyAll)()
+	rt := RestTester{DeltaSyncEnabled: base.IsEnterpriseEdition()}
```
Comment (tleyden) [inline rest/blip_api_test.go:1603]: What about introducing a `NewRestTester()` function as part of this PR which would set the `DeltaSyncEnabled` flag according to `base.IsEnterpriseEdition()`.  We wouldn't necessarily need to convert all the code over to the `NewRestTester()`, but at least new code could leverage it, and maybe at some point we could do a separate PR to refactor all the code.

File: rest/blip_sync.go [hunk 3]
```diff
@@ -800,18 +800,26 @@ func (bh *blipHandler) handleRev(rq *blip.Message) error {
...
 			return base.HTTPErrorf(http.StatusNotFound, "Can't fetch doc for deltaSrc=%s %v", deltaSrcRevID, err)
 		}
 
-		// TODO: need to add deep-copy support to revcache so deltaSrcBody doesn't get mutated
-		var deltaSrcBodyCopy map[string]interface{}
-		if err := base.DeepCopyInefficient(&deltaSrcBodyCopy, deltaSrcBody); err != nil {
-			return base.HTTPErrorf(http.StatusInternalServerError, "Error copying deltaSrcBody: %s", err)
-		}
+		// Strip out _id, _rev properties inserted by GetRevCopy
+		delete(deltaSrcBody, db.BodyId)
```
Comment (tleyden) [inline rest/blip_sync.go:819]: I wonder if this belongs in `GetRevCopy`?  It could be controlled by an extra param like `StripMetadataProps`

Since the param list is already long, it could take a `RevCopyOptions` struct as a parameter, and `StripMetadataProps` and `db.BodyDeepCopy` would be fields of that struct.

**Review by tleyden — COMMENTED** [review]
Added some suggestions for minor tweaks

**Comment by tleyden** [comment]
@adamcfraser any reason not to kick off integration tests for this?

**Comment by adamcfraser** [comment]
- [x] http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration-master/940/
- [ ] http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration-master/941/

**Review by tleyden — APPROVED** [review]
LGTM
